### PR TITLE
Rationalize Implicit Member Synthesis Somewhat

### PR DESCRIPTION
**Pull request by CodaFi**
Codable's deep magic currently forces conformance checks in the middle
of name lookup in order to inject CodingKeys into lookup results.  This
is compounded by the fact that this lookup fixup is occuring
incrementally, meaning depending on order of requirements being looked
up, IterableDeclContext::getMembers() will give you a different answer.

Compounding this, NameLookup relied on the LazyResolver to formalize
this layering violation, and relied on implicit laziness to guard
against re-entrancy.

The approach is multi-pronged:
1) Shift the layering violation into the request evaluator
2) Spell out the kinds of resolution we support explicitly (make them
easier to find and kill)
3) Remove the LazyResolver entrypoint this was relying on
4) Split off the property wrappers part into its own utility

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility Debug

**Review by DougGregor — APPROVED** [review]
Isolating the cycle-breaking hacks is incremental progress, thank you!

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -3954,6 +3954,48 @@ ConstructorDecl *NominalTypeDecl::getDefaultInitializer() const {
...
+    action.emplace(ImplicitMemberAction::ResolveImplicitInit);
+
+  if (member.isSimpleName() && !baseName.isSpecial()) {
+    if (baseName.getIdentifier() == getASTContext().Id_CodingKeys) {
+      action.emplace(ImplicitMemberAction::ResolveCodingKeys);
+    }
+  } else {
+    auto argumentNames = member.getArgumentNames();
+    if (member.isCompoundName() && argumentNames.size() != 1)
+      return;
```
Comment (CodaFi) [inline lib/AST/Decl.cpp]: This early-exit is not correct.  It's bizarre that it only makes a difference in Debug.  There's an assert somewhere that's doing lookup...

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility Debug

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility Debug

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**